packages/core/ui-settings/core-ui-settings-server-internal/src/ui_settings_config.ts

```diff
@@ -17,6 +17,12 @@ const deprecations: ConfigDeprecationProvider = ({ unused, renameFromRoot }) =>
 
 const configSchema = schema.object({
   overrides: schema.object({}, { unknowns: 'allow' }),
+  publicApiEnabled: schema.conditional(
+    schema.contextRef('serverless'),
+    true,
+    schema.boolean({ defaultValue: false }),
+    schema.never()
```

Comment:
@lukeelmers is using `schema.never` the recommended way to fully disable the option on some environment?

I don't know what other teams did, but I would have gone with something that ensure that the property is always defined on the config type, which make it less awkward when trying to access it. 

I would have replaced `schema.never` by:

```ts
schema.boolean({ 
  defaultValue: true, 
  validate: (value) => {
    if(value !== true) return "publicApiEnabled cannot be disabled"
  } 
})
```

that way, `TypeOf<typeof configSchema>`.publicApiEnabled is always present (not never-able).

Reply:
I think we briefly discussed this when @afharo was documenting the contextRef feature. `schema.never` makes access more awkward, but it also guarantees you can't possibly break things for self-managed users.

e.g. if someone were to copy `mysetting.enabled: false` to their self-managed deployment and we later removed that setting altogether, they'd get a config validation error. Even if the setting wouldn't really do anything, it would still be possible for there to be a breaking change on upgrade.

It's an edge case for sure, but I believe that's the main reason we ended up recommending `schema.never`

Reply:
Makes sense, thanks